src/doorman/test/TelnetProtocolTest.cpp

```diff
@@ -0,0 +1,270 @@
+        tp.send_telopts();
+    }
+    SECTION("Should send echo on") {
+        REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WONT, TELOPT_ECHO})));
+        tp.set_echo(true);
+    }
+    SECTION("Should send echo off") {
+        REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WILL, TELOPT_ECHO})));
+        tp.set_echo(false);
+    }
+
+    SECTION("Should handle WILLs") {
+        SECTION("handling terminal type") {
+            {
+                REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, SB, TELOPT_TTYPE, TELQUAL_SEND, IAC, SE})));
+                tp.add_data(bytes{IAC, WILL, TELOPT_TTYPE});
+            }
+            SECTION("but not the second time") { tp.add_data(bytes{IAC, WILL, TELOPT_TTYPE}); }
+        }
+        SECTION("ignoring NAWS") { tp.add_data(bytes{IAC, WILL, TELOPT_NAWS}); }
+        SECTION("refusing any option we don't understand") {
+            REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, DONT, TELOPT_BINARY})));
+            tp.add_data(bytes{IAC, WILL, TELOPT_BINARY});
+        }
+    }
+
+    SECTION("Should handle WONTs by saying DONT") {
+        REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WONT, TELOPT_BINARY})));
+        tp.add_data(bytes{IAC, DONT, TELOPT_BINARY});
+    }
+
+    SECTION("Should handle DONTs by saying WONT") {
+        REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, DONT, TELOPT_BINARY})));
+        tp.add_data(bytes{IAC, WONT, TELOPT_BINARY});
+    }
+
+    SECTION("Should handle DO") {
+        // This is confusing as heck, but passes with the original implementation.
+        SECTION("handling echo when we're echoing") {
+            REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WONT, TELOPT_ECHO})));
+            tp.add_data(bytes{IAC, DO, TELOPT_ECHO});
+        }
+        SECTION("handling echo when we're not echoing") {
+            {
+                REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WILL, TELOPT_ECHO})));
+                tp.set_echo(false);
+            }
+            REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WILL, TELOPT_ECHO})));
+            tp.add_data(bytes{IAC, DO, TELOPT_ECHO});
+        }
+        SECTION("ignoring anything else") {
+            REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WONT, TELOPT_AUTHENTICATION})));
+            tp.add_data(bytes{IAC, DO, TELOPT_AUTHENTICATION});
+        }
+    }
+
+    SECTION("Should handle sub options") {
+        SECTION("terminal types") {
+            SECTION("supporting ansi") {
+                REQUIRE_CALL(mock, on_terminal_type("xterm", true));
+                tp.add_data(bytes{IAC, SB, TELOPT_TTYPE, TELQUAL_IS, 'x', 't', 'e', 'r', 'm', IAC, SE});
+                CHECK(tp.supports_ansi());
+            }
+            SECTION("not supporting ansi") {
+                REQUIRE_CALL(mock, on_terminal_type("badger", false));
+                tp.add_data(bytes{IAC, SB, TELOPT_TTYPE, TELQUAL_IS, 'b', 'a', 'd', 'g', 'e', 'r', IAC, SE});
```

Comment:
[comment] Clearly, the badger terminal is always black and white. Looks spot on to me.